pkg/ctl/cmdutils/nodegroup_filter.go

```diff
@@ -72,70 +97,109 @@ func (f *NodeGroupFilter) ApplyExistingFilter(stackManager *manager.StackCollect
 		return err
 	}
 
-	f.existing.Insert(existing...)
+	f.ExcludeNames.Insert(existing...)
 
+	for _, name := range existing {
+		isAlsoIncluded := false
+		f.matchGlobs(name, f.includeGlobs, &isAlsoIncluded)
+		if isAlsoIncluded {
+			return fmt.Errorf("existing nodegroup %q should be excluded, but matches include fileter: %q", name, f.includeGlobsSpec)
+		}
+	}
 	return nil
 }
 
-// Match checks given nodegroup against the filter
-func (f *NodeGroupFilter) Match(ng *api.NodeGroup) bool {
-	if f.SkipAll {
-		return false
+func (*NodeGroupFilter) matchGlobs(name string, exprs []glob.Glob, result *bool) {
+	for _, compiledExpr := range exprs {
+		if compiledExpr.Match(name) {
+			*result = true
+			return
+		}
+	}
+}
+
+// Match given nodegroup against the filter and returns
+// true or false if it has to be included or excluded
+func (f *NodeGroupFilter) Match(name string) bool {
+	if f.ExcludeAll {
+		return false // force exclude
 	}
-	if f.IgnoreAllExisting && f.existing.Has(ng.Name) {
-		return false
+
+	hasIncludeRules := f.IncludeNames.Len()+len(f.includeGlobs) != 0
+	hasExcludeRules := f.ExcludeNames.Len()+len(f.excludeGlobs) != 0
+
+	if !hasIncludeRules && !hasExcludeRules {
+		return true // empty rules - include
+	}
+
+	mustInclude := false // override when rules overlap
+
+	if hasIncludeRules {
+		mustInclude = f.IncludeNames.Has(name)
+		f.matchGlobs(name, f.includeGlobs, &mustInclude)
+		if !hasExcludeRules {
+			// empty exclusion rules - strict inclusion mode
+			return mustInclude
+		}
 	}
 
-	for _, compiledExpr := range f.only {
-		if compiledExpr.Match(ng.Name) {
-			return true // return first match
+	if hasExcludeRules {
+		exclude := f.ExcludeNames.Has(name)
+		f.matchGlobs(name, f.excludeGlobs, &exclude)
+		if exclude && !mustInclude {
+			// no inclusion rules to override
+			return false
 		}
 	}
 
-	// if no globs were given, match everything,
-	// if false - we haven't matched anything so far
-	return len(f.only) == 0
+	return true // biased to include
 }
 
-// MatchAll checks all nodegroups against the filter and returns all of
-// matching names as set
-func (f *NodeGroupFilter) MatchAll(cfg *api.ClusterConfig) sets.String {
-	names := sets.NewString()
-	if f.SkipAll {
-		return names
+// MatchAll nodegroups against the filter and return two sets of names
+func (f *NodeGroupFilter) MatchAll(nodeGroups []*api.NodeGroup) (sets.String, sets.String) {
```

Comment:
We can make this signature a bit more explicit by naming the return values like this:
```func ...MatchAll(...) (included sets.String, excluded sets.String) {```
Also then we can just do `return` instead of `return included, excluded`

Also adding to the doc that the first set is the included and the second the excluded ones.

Reply:
Named returns are evil... I'll update the comment.

Reply:
Ahaaa, I had a suspicion that they could be dangerous but because I saw them in effective go I thought they were recommended. Never mind then